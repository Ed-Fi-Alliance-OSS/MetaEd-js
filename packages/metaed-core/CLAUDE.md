# CLAUDE.md

This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.

## MetaEd Core Package Overview

This is the core package of the MetaEd system, providing the compilation pipeline, entity models, and plugin infrastructure for transforming MetaEd DSL files into various output formats.

## Essential Commands

**Testing:**
- `npm test` - Run all tests for this package
- `npx jest path/to/test.ts` - Run specific test file
- `npm run test:coverage` - Run with coverage report

**Development:**
- `npm run build` - Build the package
- `npm run antlr-gen` - Regenerate ANTLR parser from grammar

## Architecture Overview

### Pipeline Flow

The `Pipeline.ts` orchestrates these sequential stages:
1. **Initialize** - Set up environment and plugins
2. **Load Files** - Read .metaed files from filesystem
3. **Validate Syntax** - Parse with ANTLR grammar
4. **Build Parse Tree** - Create ANTLR parse trees
5. **Walk Builders** - Convert parse trees to entity models
6. **Run Plugins** - Execute validators → enhancers → generators
7. **Write Output** - Save generated artifacts

### Core Components

**State Management:**
- `State.ts` - Central pipeline state containing configuration, files, entities, and results
- `MetaEdEnvironment.ts` - Runtime environment with namespaces, entities, and plugin data

**Entity System:**
- `model/` - Entity definitions (Association, DomainEntity, Common, etc.)
- `model/property/` - Property types (String, Integer, Boolean, Association, etc.)
- All entities extend `ModelBase` → `TopLevelEntity`

**Builder Pattern:**
- `builder/` - ANTLR listeners that construct entities from parse trees
- One builder per entity type (e.g., `DomainEntityBuilder`, `AssociationBuilder`)
- Two-phase processing: enter/exit methods for tree walking

**Plugin Architecture:**
- **Validators**: `(metaEd: MetaEdEnvironment) => ValidationFailure[]`
- **Enhancers**: `(metaEd: MetaEdEnvironment) => EnhancerResult`
- **Generators**: `(metaEd: MetaEdEnvironment) => Promise<GeneratorResult>`
- Plugins run in dependency order
- Validators and enhancers in each plugin run in an explicit order provided by the plugin `initialize()` function 

### Key Patterns

1. **Entity Repository**: Typed maps within namespaces for entity storage/retrieval
2. **Immutable Updates**: Never modify existing entities, create new versions
3. **Plugin Data Isolation**: Each plugin stores data under `data[pluginName]`
4. **Validation Failures**: Include source location for error reporting

### Testing Approach

Use `MetaEdTextBuilder` to construct test MetaEd code:
```typescript
MetaEdTextBuilder.build()
  .withBeginNamespace('TestNamespace')
  .withStartDomainEntity('TestEntity')
  .withStringProperty('TestProperty')
  .withEndDomainEntity()
  .sendToListener(builder);
```

### Important Files

- `index.ts` - Main exports (extensive public API)
- `Pipeline.ts` - Pipeline orchestration
- `MetaEdGrammar.g4` - ANTLR grammar definition
- `WalkBuilders.ts` - Builder orchestration logic
- `EntityRepository.ts` - Entity storage/retrieval

### Development Notes

- Grammar changes require running `npm run antlr-gen`
- New entity types need: model class, builder, and tests
- Validators should return descriptive error messages with source locations
- Enhancers must only add data, never mutate existing properties.
- Generators are async and return file content/metadata