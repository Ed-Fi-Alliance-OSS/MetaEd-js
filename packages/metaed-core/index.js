// @flow

// Test helper -- *** where should this go? ***
export { MetaEdTextBuilder } from './test/MetaEdTextBuilder';


// Functions

export { deepFreeze, deepFreezeAssign, prependIndefiniteArticle } from './src/Utility';
export { startingFromFileLoadP } from './src/task/Pipeline';
export { createMetaEdFile } from './src/task/MetaEdFile';
export { loadCoreBufferedFiles, loadExtensionBufferedFiles } from './src/task/BufferFileLoader';
export { loadFiles } from './src/task/FileSystemFilenameLoader';
export { newState } from './src/State';
export { addEntity, getAllEntities, getAllTopLevelEntities, getEntity, getTopLevelCoreEntity, getEntitiesOfType, getAllEntitiesNoSimpleTypes } from './src/model/EntityRepository';
export { addProperty, getAllProperties, getPropertiesOfType, newPropertyIndex } from './src/model/property/PropertyRepository';
export { isReferenceProperty, isSharedProperty, allPropertyTypes } from './src/model/property/PropertyType';
export { allEntityModelTypes, allTopLevelEntityModelTypes, topLevelCoreEntityModelTypes } from './src/model/ModelType';

// Builders

export { AssociationBuilder } from './src/builder/AssociationBuilder';
export { AssociationExtensionBuilder } from './src/builder/AssociationExtensionBuilder';
export { AssociationSubclassBuilder } from './src/builder/AssociationSubclassBuilder';
export { ChoiceBuilder } from './src/builder/ChoiceBuilder';
export { CommonBuilder } from './src/builder/CommonBuilder';
export { CommonExtensionBuilder } from './src/builder/CommonExtensionBuilder';
export { DecimalTypeBuilder } from './src/builder/DecimalTypeBuilder';
export { DescriptorBuilder } from './src/builder/DescriptorBuilder';
export { DomainBuilder } from './src/builder/DomainBuilder';
export { DomainEntityBuilder } from './src/builder/DomainEntityBuilder';
export { DomainEntityExtensionBuilder } from './src/builder/DomainEntityExtensionBuilder';
export { DomainEntitySubclassBuilder } from './src/builder/DomainEntitySubclassBuilder';
export { EnumerationBuilder } from './src/builder/EnumerationBuilder';
export { IntegerTypeBuilder } from './src/builder/IntegerTypeBuilder';
export { InterchangeBuilder } from './src/builder/InterchangeBuilder';
export { NamespaceInfoBuilder } from './src/builder/NamespaceInfoBuilder';
export { SharedDecimalBuilder } from './src/builder/SharedDecimalBuilder';
export { SharedIntegerBuilder } from './src/builder/SharedIntegerBuilder';
export { SharedStringBuilder } from './src/builder/SharedStringBuilder';
export { StringTypeBuilder } from './src/builder/StringTypeBuilder';

// Flow types

export type { State } from './src/State';
export type { InputDirectory } from './src/task/FileSystemFilenameLoader';
export type { FileSet, MetaEdFile } from './src/task/MetaEdFile';
export type { FileIndex } from './src/task/FileIndex';
export type { Validator } from './src/validator/Validator';
export type { ValidationFailure } from './src/validator/ValidationFailure';
export type { Enhancer } from './src/enhancer/Enhancer';
export type { EnhancerResult } from './src/enhancer/EnhancerResult';
export type { Generator } from './src/generator/Generator';
export type { GeneratorResult } from './src/generator/GeneratorResult';
export type { GeneratedOutput } from './src/generator/GeneratedOutput';
export type { MetaEdEnvironment, PluginEnvironment, SemVer } from './src/MetaEdEnvironment';
export type { MetaEdPlugin, PluginManifest } from './src/plugin/PluginTypes';

// Properties
export type { AssociationProperty } from './src/model/property/AssociationProperty';
export type { BooleanProperty } from './src/model/property/BooleanProperty';
export type { ChoiceProperty } from './src/model/property/ChoiceProperty';
export type { CommonProperty } from './src/model/property/CommonProperty';
export type { CurrencyProperty } from './src/model/property/CurrencyProperty';
export type { DateProperty } from './src/model/property/DateProperty';
export type { DecimalProperty } from './src/model/property/DecimalProperty';
export type { DescriptorProperty } from './src/model/property/DescriptorProperty';
export type { DomainEntityProperty } from './src/model/property/DomainEntityProperty';
export type { DurationProperty } from './src/model/property/DurationProperty';
export type { EntityProperty, HasReferencedEntity } from './src/model/property/EntityProperty';
export type { EnumerationProperty } from './src/model/property/EnumerationProperty';
export type { InlineCommonProperty } from './src/model/property/InlineCommonProperty';
export type { IntegerProperty } from './src/model/property/IntegerProperty';
export type { MergedProperty } from './src/model/property/MergedProperty';
export type { PercentProperty } from './src/model/property/PercentProperty';
export type { PropertyIndex } from './src/model/property/PropertyRepository';
export type { PropertyType } from './src/model/property/PropertyType';
export type { ReferentialProperty } from './src/model/property/ReferentialProperty';
export type { SchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export type { SharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export type { SharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export type { SharedShortProperty } from './src/model/property/SharedShortProperty';
export type { SharedStringProperty } from './src/model/property/SharedStringProperty';
export type { ShortProperty } from './src/model/property/ShortProperty';
export type { SimpleProperty } from './src/model/property/SimpleProperty';
export type { SharedSimple } from './src/model/SharedSimple';
export type { StringProperty } from './src/model/property/StringProperty';
export type { TimeProperty } from './src/model/property/TimeProperty';
export type { YearProperty } from './src/model/property/YearProperty';

// Entities
export type { AbstractEntity } from './src/model/AbstractEntity';
export type { Association } from './src/model/Association';
export type { AssociationExtension } from './src/model/AssociationExtension';
export type { AssociationSubclass } from './src/model/AssociationSubclass';
export type { Choice } from './src/model/Choice';
export type { Common } from './src/model/Common';
export type { CommonExtension } from './src/model/CommonExtension';
export type { DecimalType } from './src/model/DecimalType';
export type { Descriptor } from './src/model/Descriptor';
export type { Domain } from './src/model/Domain';
export type { DomainEntity } from './src/model/DomainEntity';
export type { DomainEntityExtension } from './src/model/DomainEntityExtension';
export type { DomainEntitySubclass } from './src/model/DomainEntitySubclass';
export type { DomainItem } from './src/model/DomainItem';
export type { EntityRepository } from './src/model/EntityRepository';
export type { Enumeration } from './src/model/Enumeration';
export type { EnumerationItem } from './src/model/EnumerationItem';
export type { IntegerType } from './src/model/IntegerType';
export type { Interchange } from './src/model/Interchange';
export type { InterchangeExtension } from './src/model/InterchangeExtension';
export type { InterchangeItem } from './src/model/InterchangeItem';
export type { MapTypeEnumeration } from './src/model/MapTypeEnumeration';
export type { ModelBase } from './src/model/ModelBase';
export type { ModelType } from './src/model/ModelType';
export type { NamespaceInfo } from './src/model/NamespaceInfo';
export type { SchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export type { SharedDecimal } from './src/model/SharedDecimal';
export type { SharedInteger } from './src/model/SharedInteger';
export type { SharedString } from './src/model/SharedString';
export type { SourceMap } from './src/model/SourceMap';
export type { StringType } from './src/model/StringType';
export type { Subdomain } from './src/model/Subdomain';
export type { TopLevelEntity } from './src/model/TopLevelEntity';

// Null Objects
export { NoMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { NoTopLevelEntity } from './src/model/TopLevelEntity';

// Factories
export { newMetaEdEnvironment, newPluginEnvironment } from './src/MetaEdEnvironment';
export { newMetaEdPlugin } from './src/plugin/PluginTypes';

export { newAssociationProperty } from './src/model/property/AssociationProperty';
export { newBooleanProperty } from './src/model/property/BooleanProperty';
export { newChoiceProperty } from './src/model/property/ChoiceProperty';
export { newCommonProperty } from './src/model/property/CommonProperty';
export { newCurrencyProperty } from './src/model/property/CurrencyProperty';
export { newDateProperty } from './src/model/property/DateProperty';
export { newDecimalProperty } from './src/model/property/DecimalProperty';
export { newDescriptorProperty } from './src/model/property/DescriptorProperty';
export { newDomainEntityProperty } from './src/model/property/DomainEntityProperty';
export { newDurationProperty } from './src/model/property/DurationProperty';
export { newEnumerationProperty } from './src/model/property/EnumerationProperty';
export { newInlineCommonProperty } from './src/model/property/InlineCommonProperty';
export { newIntegerProperty } from './src/model/property/IntegerProperty';
export { newMergedProperty } from './src/model/property/MergedProperty';
export { newPercentProperty } from './src/model/property/PercentProperty';
export { newSchoolYearEnumerationProperty } from './src/model/property/SchoolYearEnumerationProperty';
export { newSharedDecimalProperty } from './src/model/property/SharedDecimalProperty';
export { newSharedIntegerProperty } from './src/model/property/SharedIntegerProperty';
export { newSharedShortProperty } from './src/model/property/SharedShortProperty';
export { newSharedStringProperty } from './src/model/property/SharedStringProperty';
export { newShortProperty } from './src/model/property/ShortProperty';
export { newSimpleProperty } from './src/model/property/SimpleProperty';
export { newStringProperty } from './src/model/property/StringProperty';
export { newTimeProperty } from './src/model/property/TimeProperty';
export { newYearProperty } from './src/model/property/YearProperty';

export { newAbstractEntity } from './src/model/AbstractEntity';
export { newAssociation } from './src/model/Association';
export { newAssociationExtension } from './src/model/AssociationExtension';
export { newAssociationSubclass } from './src/model/AssociationSubclass';
export { newChoice } from './src/model/Choice';
export { newCommon } from './src/model/Common';
export { newCommonExtension } from './src/model/CommonExtension';
export { newDecimalType } from './src/model/DecimalType';
export { newDescriptor } from './src/model/Descriptor';
export { newDomain } from './src/model/Domain';
export { newDomainEntity } from './src/model/DomainEntity';
export { newDomainEntityExtension } from './src/model/DomainEntityExtension';
export { newDomainEntitySubclass } from './src/model/DomainEntitySubclass';
export { newDomainItem } from './src/model/DomainItem';
export { newEnumeration } from './src/model/Enumeration';
export { newEnumerationItem } from './src/model/EnumerationItem';
export { newInlineCommon } from './src/model/Common';
export { newIntegerType } from './src/model/IntegerType';
export { newInterchange } from './src/model/Interchange';
export { newInterchangeExtension } from './src/model/InterchangeExtension';
export { newInterchangeItem } from './src/model/InterchangeItem';
export { newMapTypeEnumeration } from './src/model/MapTypeEnumeration';
export { newNamespaceInfo } from './src/model/NamespaceInfo';
export { newSchoolYearEnumeration } from './src/model/SchoolYearEnumeration';
export { newSharedDecimal } from './src/model/SharedDecimal';
export { newSharedInteger } from './src/model/SharedInteger';
export { newSharedString } from './src/model/SharedString';
export { newSourceMap } from './src/model/SourceMap';
export { newStringType } from './src/model/StringType';
export { newSubdomain } from './src/model/Subdomain';
export { asTopLevelEntity, newTopLevelEntity } from './src/model/TopLevelEntity';
