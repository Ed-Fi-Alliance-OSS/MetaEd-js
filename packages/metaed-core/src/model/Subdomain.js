// @flow
import { newDomain } from './Domain';
import type { Domain } from './Domain';
import type { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';
import type { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import type { DomainItem } from './DomainItem';
import type { ModelBase } from './ModelBase';

export type SubdomainSourceMap = {
  ...$Exact<TopLevelEntitySourceMap>,
  domainItems: Array<SourceMap>,
  entities: Array<SourceMap>,
  parent: ?SourceMap,
  parentMetaEdName: ?SourceMap,
  position: ?SourceMap,
};

/**
 *
 */
export function newSubdomainSourceMap(): SubdomainSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    domainItems: [],
    entities: [],
    parent: NoSourceMap,
    parentMetaEdName: NoSourceMap,
    position: NoSourceMap,
  };
}

export type Subdomain = {
  sourceMap: SubdomainSourceMap,
  ...$Exact<TopLevelEntity>,
  domainItems: Array<DomainItem>,
  entities: Array<TopLevelEntity>,
  parent: Domain,
  parentMetaEdName: string,
  position: number,
};

/**
 *
 */
export function newSubdomain(): Subdomain {
  return {
    ...newTopLevelEntity(),
    type: 'subdomain',
    typeHumanizedName: 'Subdomain',
    domainItems: [],
    entities: [],
    parent: newDomain(),
    parentMetaEdName: '',
    position: 0,
    sourceMap: newSubdomainSourceMap(),
  };
}

/**
 *
 */
export const asSubdomain = (x: ModelBase): Subdomain => ((x: any): Subdomain);
