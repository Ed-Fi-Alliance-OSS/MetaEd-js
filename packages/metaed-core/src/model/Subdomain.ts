import { newDomain } from './Domain';
import { Domain } from './Domain';
import { SourceMap } from './SourceMap';
import { NoSourceMap } from './SourceMap';
import { TopLevelEntity, TopLevelEntitySourceMap } from './TopLevelEntity';
import { newTopLevelEntity, newTopLevelEntitySourceMap } from './TopLevelEntity';
import { DomainItem } from './DomainItem';
import { ModelBase } from './ModelBase';

export interface SubdomainSourceMap extends TopLevelEntitySourceMap {
  domainItems: Array<SourceMap>;
  entities: Array<SourceMap>;
  parent: SourceMap;
  parentMetaEdName: SourceMap;
  position: SourceMap;
}

/**
 *
 */
export function newSubdomainSourceMap(): SubdomainSourceMap {
  return {
    ...newTopLevelEntitySourceMap(),
    domainItems: [],
    entities: [],
    parent: NoSourceMap,
    parentMetaEdName: NoSourceMap,
    position: NoSourceMap,
  };
}

export interface Subdomain extends TopLevelEntity {
  sourceMap: SubdomainSourceMap;
  domainItems: Array<DomainItem>;
  entities: Array<TopLevelEntity>;
  parent: Domain;
  parentMetaEdName: string;
  position: number;
}

/**
 *
 */
export function newSubdomain(): Subdomain {
  return {
    ...newTopLevelEntity(),
    type: 'subdomain',
    typeHumanizedName: 'Subdomain',
    domainItems: [],
    entities: [],
    parent: newDomain(),
    parentMetaEdName: '',
    position: 0,
    sourceMap: newSubdomainSourceMap(),
  };
}

/**
 *
 */
export const asSubdomain = (x: ModelBase): Subdomain => x as Subdomain;
