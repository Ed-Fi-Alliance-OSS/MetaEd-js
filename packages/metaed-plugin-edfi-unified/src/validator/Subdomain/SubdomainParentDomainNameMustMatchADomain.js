// @flow
import type { Subdomain, MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';
import type { SubdomainSourceMap } from '../../../../../packages/metaed-core/src/model/Subdomain';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.subdomain.forEach((subdomain: Subdomain) => {
    if (!metaEd.entity.domain.has(subdomain.parentMetaEdName)) {
      failures.push({
        validatorName: 'SubdomainParentDomainNameMustMatchADomain',
        category: 'error',
        message: `Subdomain parent domain name '${subdomain.parentMetaEdName}' does not match any declared Domain.`,
        sourceMap: ((subdomain.sourceMap: any): SubdomainSourceMap).parentMetaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
