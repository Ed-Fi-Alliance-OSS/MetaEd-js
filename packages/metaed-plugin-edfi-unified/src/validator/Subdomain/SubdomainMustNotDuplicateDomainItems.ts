import { MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { findDuplicates } from '../ValidatorShared/FindDuplicates';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.subdomain.forEach(subdomain => {
      const names = subdomain.domainItems.map(di => di.metaEdName);
      const duplicates: Array<string> = findDuplicates(names);

      duplicates.forEach(val => {
        const domainItem = subdomain.domainItems.find(d => d.metaEdName === val);
        if (domainItem !== undefined) {
          failures.push({
            validatorName: 'SubdomainMustNotDuplicateDomainItems',
            category: 'error',
            message: `Subdomain Entity Domain Item '${domainItem.metaEdName}' has a duplicate Domain Item.`,
            sourceMap: domainItem.sourceMap.type,
            fileMap: null,
          });
        }
      });
    });
  });

  return failures;
}
