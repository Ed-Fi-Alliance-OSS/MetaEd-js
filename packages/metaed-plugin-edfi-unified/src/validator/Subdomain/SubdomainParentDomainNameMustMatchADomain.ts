import { Subdomain, SubdomainSourceMap, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.subdomain.forEach((subdomain: Subdomain) => {
      if (getEntityForNamespaces(subdomain.parentMetaEdName, [namespace, ...namespace.dependencies], 'domain') == null) {
        failures.push({
          validatorName: 'SubdomainParentDomainNameMustMatchADomain',
          category: 'error',
          message: `Subdomain parent domain name '${subdomain.parentMetaEdName}' does not match any declared Domain.`,
          sourceMap: (subdomain.sourceMap as SubdomainSourceMap).parentMetaEdName,
          fileMap: null,
        });
      }
    });
  });
  return failures;
}
