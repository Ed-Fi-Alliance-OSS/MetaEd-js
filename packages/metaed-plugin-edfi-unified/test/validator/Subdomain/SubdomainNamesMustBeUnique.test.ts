import { newMetaEdEnvironment, MetaEdTextBuilder, DomainBuilder, NamespaceBuilder } from 'metaed-core';
import { MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { validate } from '../../../src/validator/Subdomain/SubdomainNamesMustBeUnique';

describe('when entities in same namespace have different names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()

      .withStartSubdomain('ValidName2', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build two subdomain', () => {
    expect(coreNamespace.entity.subdomain.size).toBe(2);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when entities in same namespace have identical names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()

      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one subdomain because DomainBuilder will not let it get that far', () => {
    expect(coreNamespace.entity.subdomain.size).toBe(1);
  });

  it('should have no validation failures because of DomainBuilder', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when subdomains in separate dependency-linked namespaces have identical names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;
  let extensionNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .withBeginNamespace('Extension', 'Extension')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespace = metaEd.namespace.get('Extension');
    extensionNamespace.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should build one subdomain in core namespace', () => {
    expect(coreNamespace.entity.subdomain.size).toBe(1);
  });

  it('should build one subdomain in extension namespace', () => {
    expect(extensionNamespace.entity.subdomain.size).toBe(1);
  });

  it('should have validation failures for each entity', () => {
    expect(failures).toHaveLength(2);

    expect(failures[0].validatorName).toBe('SubdomainNamesMustBeUnique');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();

    expect(failures[1].validatorName).toBe('SubdomainNamesMustBeUnique');
    expect(failures[1].category).toBe('error');
    expect(failures[1].message).toMatchSnapshot();
    expect(failures[1].sourceMap).toMatchSnapshot();
  });
});

describe('when subdomains in non-dependency-linked namespaces have identical names', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  let failures: Array<ValidationFailure>;
  let coreNamespace: any = null;
  let extensionNamespacea: any = null;
  let extensionNamespaceb: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSubdomain('NotADuplicate', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .withBeginNamespace('Extensiona', 'Extensiona')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .withBeginNamespace('Extensionb', 'Extensionb')
      .withStartSubdomain('ValidName1', 'DomainName')
      .withDocumentation('doc')
      .withDomainEntityDomainItem('SubdomainItem1')
      .withEndSubdomain()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    extensionNamespacea = metaEd.namespace.get('Extensiona');
    extensionNamespaceb = metaEd.namespace.get('Extensionb');
    // $FlowIgnore - null check
    extensionNamespacea.dependencies.push(coreNamespace);
    // $FlowIgnore - null check
    extensionNamespaceb.dependencies.push(coreNamespace);

    failures = validate(metaEd);
  });

  it('should build one core subdomain', () => {
    expect(coreNamespace.entity.subdomain.size).toBe(1);
  });

  it('should build one extension1 subdomain', () => {
    expect(extensionNamespacea.entity.subdomain.size).toBe(1);
  });

  it('should build one extension2 subdomain', () => {
    expect(extensionNamespaceb.entity.subdomain.size).toBe(1);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});
