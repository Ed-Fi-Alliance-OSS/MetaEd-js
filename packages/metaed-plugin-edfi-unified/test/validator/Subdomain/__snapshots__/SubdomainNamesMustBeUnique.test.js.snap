// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when subdomains in separate dependency-linked namespaces have identical names should have validation failures for each entity 1`] = `"Subdomain named ValidName1 is a duplicate declaration of that name."`;

exports[`when subdomains in separate dependency-linked namespaces have identical names should have validation failures for each entity 2`] = `
Object {
  "column": 2,
  "line": 8,
  "tokenText": "Subdomain",
}
`;

exports[`when subdomains in separate dependency-linked namespaces have identical names should have validation failures for each entity 3`] = `"Subdomain named ValidName1 is a duplicate declaration of that name."`;

exports[`when subdomains in separate dependency-linked namespaces have identical names should have validation failures for each entity 4`] = `
Object {
  "column": 2,
  "line": 2,
  "tokenText": "Subdomain",
}
`;
